libvast/src/plugin.cpp

```diff
@@ -380,6 +380,38 @@ store_plugin::make_store(system::accountant_actor accountant,
                                                  std::move(path), name());
 }
 
+// -- loader plugin -------------------------------------------------------------
+
+auto stdin_loader_plugin::make_loader(options, const bool_operator*) const
+  -> caf::expected<input_loader> {
+  return []() -> detail::generator<chunk_ptr> {
+    auto input = std::string{};
+    auto c = char{};
+    while (std::cin.get(c)) {
+      input += c;
+    }
+    if (not input.empty()) {
+      auto chunk = chunk::make(std::move(input));
+      co_yield std::move(chunk);
+    }
+    co_return;
```

Comment:
I think there are some problems here that we need to take a look at with some detail:
- When do we yield? We don't want to read everything and then yield just a single chunk, as that destroys the pipelining we're trying to achieve
- How do we handle a stalling input? `std::istream::get(char&)` may block indefinitely.
- Can we read more than one character at a time? Maybe BUFSIZ is a good starting point, which you could use through `rdbuf`? (Although there is a [blog post](https://www.zverovich.net/2020/08/04/optimal-file-buffer-size.html) from a well-known author saying that the value should be larger than that, at least for writing.)
- Do we need to go through the iostreams API for this?

Reply:
For now, I'd go with a reasonably large buffer size like 16k. At some later point, I'd then do some benchmarks. Our `fdinbuf` has half the size as default: https://github.com/tenzir/vast/blob/master/libvast/include/vast/detail/fdinbuf.hpp

Reply:
Looking at fdinbuf we can probably get rather far with it. Any logic for incremental backoff should likely be implemented generically upstream of the source in a pipeline, so in case of a timeout it should suffice to emit an empty chunk.

Reply:
> Looking at fdinbuf we can probably get rather far with it.

Yeah, we use it for quite a while now when reading from STDIN, so it's battle-proven.

> Any logic for incremental backoff should likely be implemented generically upstream of the source in a pipeline, so in case of a timeout it should suffice to emit an empty chunk.

What's the benefit of emitting empty chunks? Why simply not emit anything when hitting a timeout?

Reply:
You need to yield control to outside of the generator coroutine, for which you must yield something. If you have nothing but are not at the end of your input, then you yield nothing, which in this case is an empty chunk. The executor does not forward empty batches of elements (empty chunks or empty table slices) to the next operator.

Reply:
`fdinbuf` is what we are looking for. I think we can even omit the char-by-char insertion from the buffer into a string but I might need to implement some kind of buffer slice span method for `fdinbuf`.

Reply:
There are two ways to go about this after reading something:
1. You can swap out the rdbuf and "move" the ownership of the old rdbuf into the chunk
2. You can copy from the rdbuf into a newly created chunk

For (1) you'll likely need a way to access an rvalue reference to `fdinbuf::buffer_`, and for (2) you'll likely need an `as_bytes(const fdinbuf&)` overload. Both should work; I'm not sure without measuring how the extra copy in (2) compares against the fdinbuf construction and rdbuf swap in (1). I wouldn't recommend spending time measuring that, but please add a comment to the code explaining that we have these two options and did not benchmark.

Reply:
Hmmm, I think I've been too naive with `fdinbuf::buffer_`... I noticed that `fdinbuf` uses a putback area in its `buffer` with a `putback_area_size` of 10 - e.g. when beginning to read data, `fdinbuf`  only starts writing into `buffer_` from its 10th index. Then, in the next chunk, the putback area will be then filled with what looks like `gptr() - putback_area_size`. So basically I can not simply use all the contents of the `buffer_` and an `{eback(), egptr()}` slice of the `buffer_`  will always return a chunk with a maximum size of `~16k - putback_area_size`.

The main directions I can go to from here are:
1. Continue the `{eback(), egptr()}` approach.
2. Check whether `fdinbuf` actually needs a putback area in our case.
3. Disregard direct-from-buffer approach and append a (size-reserved) string char-by-char.

Reply:
As a variation (3b), you could take some inspiration from `detail::absorb_line`.

Reply:
I decided to opt for the 3-3b implementation.